ANPR/src/anpr/license_plate_recognition.py

```diff
@@ -0,0 +1,283 @@
+        self.debug_print(f'Picking out of {len(candidates)} candidates.', 2)
+        for c in candidates:
+            # compute the bounding box of the contour and then use
+            # the bounding box to derive the aspect ratio
+            (x, y, w, h) = cv2.boundingRect(c)
+            location = ResultLocation((x, y), w, h)
+            ar = w / float(h)
+
+            # check to see if the aspect ratio is rectangular
+            if self.minAR <= ar <= self.maxAR:
+                # store the license plate contour and extract the
+                # license plate from the grayscale image and then
+                # threshold it
+                licensePlate = gray[y:y + h, x:x + w]
+                roi_temp = cv2.threshold(licensePlate, 0, 255,
+                                         cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]
+
+                # Select the license plate with the longest text on it
+                ocr_results = self.selection_ocr.getTextFromImage(roi_temp)  # Read text
+                self.debug_print(f"Found Text on Candidate: {[r.text for r in ocr_results]}", 3)
+                if len(ocr_results) > 0:
+                    self.debug_imshow("License Plate Candidate", licensePlate, 3)
+                    self.debug_imshow("ROI", roi_temp, 3, waitKey=True, )
+                    text_length = sum(len(r.text) for r in ocr_results)
+                    if text_length is None or longest_text_length < text_length:
+                        longest_text_length = text_length
+                        lp_location = location
+                        roi = roi_temp
+
+        # return a 2-tuple of the license plate ROI and the contour
+        # associated with it
+        if roi is not None:
+            cv2.imshow("ROI", roi)
+        return roi, lp_location
+
+    def find_and_ocr(self, image, doSelection=True):
+        # convert the input image to grayscale, locate all candidate
+        # license plate regions in the image, and then process the
+        # candidates, leaving us with the *actual* license plate
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+
+        location = None
+        if doSelection:
+            candidates = self.locate_license_plate_candidates(gray)
+            (lp_image, location) = self.pick_license_plate_location(gray, candidates)
+        else:
+            lp_image = gray
+
+        # only OCR the license plate if the license plate ROI is not
+        # empty
+        if lp_image is not None:
+            # OCR the license plate
+            ocr_results = self.result_ocr.getTextFromImage(lp_image)
+
+            lp_results = [LicensePlateResult(result.text, result.location) for result in ocr_results]
+
+            if self.formats is not None:
+                formatted_results = []
+                for result in lp_results:
+                    new_results = result.match_format(self.formats)
+                    formatted_results += new_results
```

Comment:
```suggestion
formatted_results = lp_results.map(lambda result: result.match_format(self.formats)
```

Reply:
You can choose to use a list comprehension or map (the latter is most often faster).